notify/sns/sns.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+	commoncfg "github.com/prometheus/common/config"
+)
+
+// Notifier implements a Notifier for SNS notifications.
+type Notifier struct {
+	conf    *config.SNSConfig
+	tmpl    *template.Template
+	logger  log.Logger
+	client  *http.Client
+	retrier *notify.Retrier
+}
+
+func (n Notifier) Notify(ctx context.Context, alert ...*types.Alert) (bool, error) {
+	credentials := credentials.NewStaticCredentials(n.conf.Sigv4.AccessKey, string(n.conf.Sigv4.SecretKey), "")
+	if n.conf.Sigv4.AccessKey == "" {
+		credentials = nil
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Region:      aws.String(n.conf.Sigv4.Region),
+			Credentials: credentials,
+			Endpoint:    aws.String(n.conf.APIUrl),
+		},
+		Profile: n.conf.Sigv4.Profile,
+	})
+
+	if _, err := sess.Config.Credentials.Get(); err != nil {
+		return false, fmt.Errorf("could not get SigV4 credentials: %w", err)
+	}
+
+	if n.conf.Sigv4.RoleARN != "" {
+		sess.Config.Credentials = stscreds.NewCredentials(sess, n.conf.Sigv4.RoleARN)
+	}
+
+	data := notify.GetTemplateData(ctx, n.tmpl, alert, n.logger)
+	tmpl := notify.TmplText(n.tmpl, data, &err)
+	message := tmpl(n.conf.Message)
+
+	client := sns.New(sess, &aws.Config{Credentials: credentials})
+	publishInput := &sns.PublishInput{}
+
+	if n.conf.TopicARN != "" {
+		publishInput.SetTopicArn(n.conf.TopicARN)
+		messageToSend, isTrunc, err := validateAndTruncateMessage(message)
+		if err != nil {
+			return false, err
+		}
+		if isTrunc {
+			n.conf.Attributes["truncated"] = "true"
+		}
+		publishInput.SetMessage(messageToSend)
+	}
+	if n.conf.PhoneNumber != "" {
+		publishInput.SetPhoneNumber(n.conf.PhoneNumber)
+		// If SMS message is over 1600 chars, SNS will reject the message.
+		_, isTruncated := notify.Truncate(message, 1600)
+		if isTruncated {
+			return false, fmt.Errorf("SMS message exeeds length of 1600 charactors")
+		} else {
+			publishInput.SetMessage(message)
+		}
+	}
+	if n.conf.TargetARN != "" {
+		publishInput.SetTargetArn(n.conf.TargetARN)
+		messageToSend, isTrunc, err := validateAndTruncateMessage(message)
+		if err != nil {
+			return false, err
+		}
+		if isTrunc {
+			n.conf.Attributes["truncated"] = "true"
+		}
+		publishInput.SetMessage(messageToSend)
+	}
+
+	if len(n.conf.Attributes) > 0 {
+		attributes := map[string]*sns.MessageAttributeValue{}
+		for k, v := range n.conf.Attributes {
+			attributes[k] = &sns.MessageAttributeValue{DataType: aws.String("String"), StringValue: aws.String(v)}
+		}
+		publishInput.SetMessageAttributes(attributes)
+	}
+
+	if n.conf.Subject != "" {
+		publishInput.SetSubject(n.conf.Subject)
+	}
+
+	// Deduplication key is only added if it's a FIFO SNS Topic.
+	if n.conf.IsFIFOTopic {
+		key, err := notify.ExtractGroupKey(ctx)
+		if err != nil {
+			return false, err
+		}
+		publishInput.SetMessageDeduplicationId(key.Hash())
```

Comment:
I see that's part of the original specifications but I'm not sure I fully understand why we set it to the group key's hash. To my understanding (but I may be wrong) is that the 5-minutes deduplication time window done by SNS will conflict with the Alertmanager `group_interval` if it's set less to 5 minutes because we may loose new alerts added to the group because of the SNS deduplication.

Reply:
Yeah, I confirmed the group key hash isn't unique enough to handle a message with the same title but different labels send in the 5 minute SNS de-duplication window. I'll ask in the issue, we're also missing a message group id for FIFO queues that is required by AWS.